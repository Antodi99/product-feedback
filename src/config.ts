export const BACKEND_API_URL = 'https://go-product-feedback.herokuapp.com'
