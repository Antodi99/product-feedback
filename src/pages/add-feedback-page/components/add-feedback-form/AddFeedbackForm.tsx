import { Formik } from 'formik'
import { Link, useNavigate } from 'react-router-dom'
import { addFeedback } from '../../../../services/feedback.service'
import { Dropdown } from '../../../../components/Dropdown'
import { optionsCategory } from './constants'
import { addFeedbackSchema } from './validation'

type FormValues = {
  detail: string
  title: string
  category: string
}

const initialValues: FormValues = {
  title: '',
  detail: '',
  category: optionsCategory[0],
}

export function AddFeedback() {
  const navigate = useNavigate()

  const handleAddFeedback = async ({ detail, title, category }: FormValues) => {
    await addFeedback(detail, title, category)
    navigate('/')
  }

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={(values) => {
        handleAddFeedback(values)
      }}
      validationSchema={addFeedbackSchema}
      validateOnBlur
    >
      {({
        values,
        handleChange,
        handleSubmit,
        errors,
        touched,
        handleBlur,
      }) => {
        return (
          <div className='bg-white flex flex-col rounded-lg p-6 h-fit w-full'>
            <h1 className='text-dark-blue text-2xl font-bold'>
              Create New Feedback
            </h1>
            <div className='mt-4'>
              <p className='text-dark-blue text-sm font-bold'>Feedback Title</p>
              <p className='text-dark-blue text-sm'>
                Add a short, descriptive headline
              </p>
              <input
                className={
                  errors.title && touched.title
                    ? 'bg-light-grey p-2 w-full outline-none rounded-lg mt-4 border-rose-500 border-2 border-solid'
                    : 'bg-light-grey p-2 w-full outline-none rounded-lg mt-4'
                }
                onChange={handleChange}
                value={values.title}
                id='title'
                name='title'
                onBlur={handleBlur}
              ></input>
              {errors.title && touched.title ? (
                <div className='text-rose-500 mt-1'>{errors.title}</div>
              ) : null}
            </div>
            <div className='mt-8'>
              <p className='text-dark-blue text-sm font-bold'>Category</p>
              <p className='text-dark-blue text-sm'>
                Choose a category for you feedback
              </p>
              <Dropdown
                selected={values.category}
                handleChange={handleChange as any}
                options={optionsCategory}
                errors={errors.category}
                touched={touched.category}
              />
              {errors.category && touched.category ? (
                <div className='text-rose-500'>{errors.category}</div>
              ) : null}
            </div>
            <div className='mt-8'>
              <p className='text-dark-blue text-sm font-bold'>
                Feedback Detail
              </p>
              <p className='text-dark-blue text-sm'>
                Include any specific comments on what should be improved, added,
                etc.
              </p>
              <textarea
                className={
                  errors.detail && touched.detail
                    ? 'bg-light-grey p-2 resize-none w-full outline-none rounded-lg mt-4 h-32 border-rose-500 border-2 border-solid'
                    : 'bg-light-grey p-2 resize-none w-full outline-none rounded-lg mt-4 h-32'
                }
                onChange={handleChange}
                value={values.detail}
                name='detail'
                id='detail'
                onBlur={handleBlur}
              ></textarea>
              {errors.detail && touched.detail ? (
                <div className='text-rose-500'>{errors.detail}</div>
              ) : null}
            </div>
            <div className='flex flex-col-reverse md:flex-row md:justify-end mt-10'>
              <Link to={'/'}>
                <div className='rounded-xl p-2 md:p-3 mt-3 md:mt-0 md:px-6 bg-dark-blue w-content flex items-center justify-center hover:bg-light-blue cursor-pointer'>
                  <p className='text-white font-bold text-sm'>Cancel</p>
                </div>
              </Link>
              <div
                className='rounded-xl p-2 md:p-3 bg-fuchsia-500 w-content md:ml-3 flex items-center justify-center hover:bg-fuchsia-400 cursor-pointer'
                onClick={() => handleSubmit()}
              >
                <p className='text-white font-bold text-sm'>Add Feedback</p>
              </div>
            </div>
          </div>
        )
      }}
    </Formik>
  )
}
