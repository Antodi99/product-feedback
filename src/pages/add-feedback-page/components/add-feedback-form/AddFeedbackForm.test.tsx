import nock from 'nock'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { AddFeedback } from './AddFeedbackForm'
import { BACKEND_API_URL } from '../../../../config'
import { MemoryRouter, Route, Routes } from 'react-router-dom'

describe('AddFeedback', () => {
  const defaultFeedbackRoute = `/feedback/add`

  const setup = (url = defaultFeedbackRoute) => {
    return render(
      <MemoryRouter initialEntries={[url]}>
        <Routes>
          <Route path='/' element={<div>Home</div>}></Route>
          <Route path='/feedback/add' element={<AddFeedback />}></Route>
        </Routes>
      </MemoryRouter>
    )
  }

  it('renders properly', async () => {
    const { container } = setup()
    expect(container).toMatchSnapshot()
  })

  it('renders properly', async () => {
    const scope = nock(BACKEND_API_URL)
      .post(`/api/feedback/`)
      .matchHeader('authorization', /^Bearer .*$/)
      .times(1)
      .reply(200)

    setup()

    const title = await screen.findByTestId('title')
    await userEvent.type(title, 'Add Comment')

    const detail = screen.getByTestId('detail')
    await userEvent.type(detail, 'Add Comment')

    await userEvent.click(screen.getByText('Add Feedback'))
    scope.done()
    await screen.findByText(/home/i)
  })

  it('user clicks cancel button', async () => {
    setup()

    const cancelBtn = await screen.findByText(/cancel/i)
    userEvent.click(cancelBtn)
    await screen.findByText(/home/i)
  })
})
