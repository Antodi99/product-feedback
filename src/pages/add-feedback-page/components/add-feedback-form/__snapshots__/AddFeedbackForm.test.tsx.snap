// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFeedback renders properly 1`] = `
<div>
  <div
    class="bg-white flex flex-col rounded-lg p-6 h-fit w-full"
  >
    <h1
      class="text-dark-blue text-2xl font-bold"
    >
      Create New Feedback
    </h1>
    <div
      class="mt-4"
    >
      <p
        class="text-dark-blue text-sm font-bold"
      >
        Feedback Title
      </p>
      <p
        class="text-dark-blue text-sm"
      >
        Add a short, descriptive headline
      </p>
      <input
        class="bg-light-grey p-2 w-full outline-none rounded-lg mt-4"
        data-testid="title"
        id="title"
        name="title"
        value=""
      />
    </div>
    <div
      class="mt-8"
    >
      <p
        class="text-dark-blue text-sm font-bold"
      >
        Category
      </p>
      <p
        class="text-dark-blue text-sm"
      >
        Choose a category for you feedback
      </p>
      <div
        class="relative inline-block text-left w-full bg-light-grey mt-4 rounded-lg"
      >
        <div>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            class="-z-10 inline-flex w-full rounded-md px-4 py-2 font-medium text-light-blue hover:text-fuchsia-400 justify-between"
            data-testid="category"
            id="headlessui-menu-button-:r0:"
            type="button"
          >
            UI
            <svg
              aria-hidden="true"
              class="-mr-1 h-5 w-5 ease-linear duration-150 text-light-blue"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="mt-8"
    >
      <p
        class="text-dark-blue text-sm font-bold"
      >
        Feedback Detail
      </p>
      <p
        class="text-dark-blue text-sm"
      >
        Include any specific comments on what should be improved, added, etc.
      </p>
      <textarea
        class="bg-light-grey p-2 resize-none w-full outline-none rounded-lg mt-4 h-32"
        data-testid="detail"
        id="detail"
        name="detail"
      />
    </div>
    <div
      class="flex flex-col-reverse md:flex-row md:justify-end mt-10"
    >
      <a
        href="/"
      >
        <div
          class="rounded-xl p-2 md:p-3 mt-3 md:mt-0 md:px-6 bg-dark-blue w-content flex items-center justify-center hover:bg-light-blue cursor-pointer"
        >
          <p
            class="text-white font-bold text-sm"
          >
            Cancel
          </p>
        </div>
      </a>
      <div
        class="rounded-xl p-2 md:p-3 bg-fuchsia-500 w-content md:ml-3 flex items-center justify-center hover:bg-fuchsia-400 cursor-pointer"
        data-testid="addFeedback"
      >
        <p
          class="text-white font-bold text-sm"
        >
          Add Feedback
        </p>
      </div>
    </div>
  </div>
</div>
`;
